File: Users.API/Users.Logic/Services/UserService.cs
using Azure;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Users.Data;
using Users.Data.Models;
using Users.Logic.Models.Requests;
using Users.Logic.Models.Responses;

namespace Users.Logic.Services
{
    public class UserService : IUserService
    {
        private readonly MyDBContext _dbContext;

        public UserService(MyDBContext dbContext)
        {
            _dbContext = dbContext;
        }

        public async Task<AddUserResponse> AddUser(AddUserRequest addUserRequest)
        {
            try
            {
                User? user = Transformers.UserTransform.CreateObject(addUserRequest.User);

                if (user != null)
                {
                    await _dbContext.Users.AddAsync(user);
                    await _dbContext.SaveChangesAsync();
                }

                return new AddUserResponse()
                {
                    User = Transformers.UserTransform.CreateObject(user)
                };
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<GetUserResponse> GetUsers(long? id, string? name, string? surname)
        {
            try
            {
                IQueryable<User> userQueryable = from user in _dbContext.Users select user;

                if (id.HasValue && id.Value > 0)
                    userQueryable = userQueryable.Where(u => u.Id == id);

                if (!string.IsNullOrEmpty(name))
                    userQueryable = userQueryable.Where(u => u.Nome.Contains(name));

                if (!string.IsNullOrEmpty(surname))
                    userQueryable = userQueryable.Where(u => u.Sobrenome.Contains(surname));

                User[] users = await userQueryable.ToArrayAsync();

                return new GetUserResponse()
                {
                    Users = users.Select(u => Transformers.UserTransform.CreateObject(u)!).ToArray()
            };
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<UpdateUserResponse> UpdateUser(UpdateUserRequest updateUserRequest)
        {
            try
            {
                if (updateUserRequest.User == null)
                    throw new Exception("Invalid user");

                User? user = await _dbContext.Users.Where(u => u.Id == updateUserRequest.User.Id).FirstOrDefaultAsync();

                if (user == null)
                    throw new Exception("User not found");

                Transformers.UserTransform.UpdateObject(ref user, updateUserRequest.User);

                if (user != null && _dbContext.ChangeTracker.HasChanges())
                    _dbContext.SaveChanges();

                return new UpdateUserResponse()
                {
                    User = Transformers.UserTransform.CreateObject(user)
                };
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task<RemoveUserResponse> DeleteUser(RemoveUserRequest removeUserRequest)
        {
            try
            {
                if (removeUserRequest.User == null)
                    throw new Exception("Invalid user");

                User? user = await _dbContext.Users.Where(u => u.Id == removeUserRequest.User.Id).FirstOrDefaultAsync();

                if (user == null)
                    throw new Exception("User not found");

                _dbContext.Users.Remove(user);
                _dbContext.SaveChanges();

                return new RemoveUserResponse()
                {
                    User = Transformers.UserTransform.CreateObject(user)
                };
            }
            catch (Exception)
            {
                throw;
            }
        }
    }
}

File: Users.UI/src/app/@core/services/user.service.ts
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { AddUserRequest, AddUserResponse, GetUserResponse, RemoveUserRequest, RemoveUserResponse, UpdateUserRequest, UpdateUserResponse, UsersClient } from "../http/users-client";

@Injectable()
export class UserService{
    constructor(private userClient: UsersClient){}

    getUsers(id: number | undefined, name: string | undefined, surname: string | undefined) : Observable<GetUserResponse>{
        const getUsers$: Observable<GetUserResponse> = this.userClient.userGET(id, name, surname);
        return getUsers$;
    }

    addUser(addUserRequest: AddUserRequest) : Observable<AddUserResponse>{
        const addUser$: Observable<AddUserResponse> = this.userClient.userPOST(addUserRequest);
        return addUser$;
    }

    updateUser(updateUserRequest: UpdateUserRequest) : Observable<UpdateUserResponse>{
        const updateUser$: Observable<UpdateUserResponse> = this.userClient.userPUT(updateUserRequest);
        return updateUser$;
    }

    removeUser(removeUserRequest: RemoveUserRequest): Observable<RemoveUserResponse>{
        const deleteUser$: Observable<RemoveUserResponse> = this.userClient.userDELETE(removeUserRequest);
        return deleteUser$;
    }
}